### ISSUE TITLE
Manually removing ties in dynamic simulation

### ISSUE THREAD

**Issue by ThomasKraft** [statnet/EpiModel#494]
I am working with a full set of custom modules in which individuals have a dynamic attribute "zed" that defines which nodes are eligible to interact in the network (ties only possible if nodes have same "zed" value). My underlying TERGM makes it impossible for ties to form between individuals that do not share the same value of "zed". However, individuals change "zed" dynamically as the model runs, and thus I have a need to dissolve ties manually for individuals as this changes (e.g. nodes 1 and 2 share a tie and both have zed "5", but then node 1 moves to zed "6" and thus their existing tie must immediately end).

I have come up with two solutions that I embed within a custom module to manually dissolve such ties early:
```
# APPROACH 1
nw.df <- as.data.frame(dat$nw)
nw.df <- nw.df[which(nw.df$onset < at),]
nw.df$nw.df_tail_zed <- get_attr(dat, "zed")[nw.df$tail]
nw.df$nw.df_head_zed <- get_attr(dat, "zed")[nw.df$head]
heteros <- nw.df[nw.df$nw.df_head_zed != nw.df$nw.df_tail_zed,]
dat$nw[[1]] <- deactivate.edges(dat$nw[[1]], onset = at, terminus=Inf, e = heteros$edge.id)
```
And
```
# APPROACH 2
moved # vector of ids indices that should have ties broken
moved_edges <- as.numeric( unlist(sapply(moved, function(x) get.edgeIDs.active(dat$nw[[1]], v = x , at = at ))) )

if(length(moved_edges) > 0 ){
  dat$nw[[1]] <- deactivate.edges(x = dat$nw[[1]], onset=at, terminus=at, e = moved_edges)
}
```

I have three questions that I am struggling with:
1) Is it immediately obvious why approach 1 above works, but approach 2 does not? I would generally prefer not to use`as.data.frame(dat$nw)` directly within a module and suspect it's bad practice.
2) Even though approach 1 works, is there a better way to accomplish this?
3) I am trying to make my existing module compatible with `tergmLite`, but these methods require accessing `dat$nw[[1]]` directly. This causes an error because the network object is not being saved, and so I think I would need to perform an operation on the edgelist object that is used instead. Any advice on how I might translate this? (perhaps making the edgelist a dataframe and adding the "zed" attribute for egos and alters as separate columns, and then removing edgelist rows where those are not the same?)

**Comment by smjenness** [comment]
1. I am not sure why the first approach might work but the second does not, without stepping through your code. Perhaps @AdrienLeGuillou has some general thoughts here. 
2. I assume you do not have a mixing term in your dissolution model (e.g., a nodemix with the off-diagonal duration set to 1.01 [have to set to >1 for structural reasons). Check out the conversation on https://github.com/statnet/EpiModel/issues/483, as it may be related.
3. tergmLite transforms the network class object to a two-column edgelist, which is what your first line in the approach 1 is doing. You can extract that with `nw.df <- dat$el[[1]]` instead, and then perform the subsequent operations.

**Comment by AdrienLeGuillou** [comment]
I understand how APPROACH 1 works but APPROACH 2 is less clear.

I think your are trying to get the EdgeIDs of the heterogenous pairs. But from what I understand `get.edgeIDs.active` gives you all the edges implying vertex `v` at time `at`. 

From what @smjenness said and what I understand you want to do, here is something that should work. @smjenness I would need your input on that as the EdgeList part is not my forte.

``` R
# get the "zed" attribute vector and the edgelist
zed <- get_attr(dat, "zed")
el <- dat$el[[1]]

# Find the different zed values among the ties
heteros <- zed[el[, 1]] != zed[el[, 2]]

# Remove them from the edgelist
el <- el[!heteros, ]

# Update the "dat" edgelist
dat$el[[1]] <- updated_el
```

Also quick note:

```R
as.numeric( unlist(sapply(moved, function(x) get.edgeIDs.active(dat$nw[[1]], v = x , at = at ))) )
```

should be written with `vapply` (safe version of `lapply`):

```R
# with explicitly named parameters for pedagogical purposes
vapply(
  X = moved, 
  FUN = function(x) get.edgeIDs.active(dat$nw[[1]], v = x , at = at), 
  FUN.VALUE = numeric(1)
)
```

Generally, `sapply` should be banned from "development code" (but ok in analysis scripts)

**Comment by ThomasKraft** [comment]
Thank you for the insights.

> 2. I assume you do not have a mixing term in your dissolution model (e.g., a nodemix with the off-diagonal duration set to 1.01 [have to set to >1 for structural reasons). Check out the conversation on https://github.com/[/issues/483](https://github.com/statnet/EpiModel/issues/483), as it may be related.

@smjenness I haven't included a `nodemix` term in my dissolution model because I don't have a `nodemix` in my formation model (although I believe you have been working on relaxing that constraint?). Interestingly, the reason for this approach is that I was addressing the same issue that you discuss in the thread for #483 by instead implementing a block diagonal constraint using a method suggested by @chad-klumb (https://github.com/statnet/EpiModel/issues/435#issuecomment-666525487). Both cases are dealing with migration and no desired ties between groups. Is there any chance you could comment on the relative merit of these approaches (offset with -Inf coef in formation and hard-coded removal of ties on migration VS. nodemix terms in formation and dissolution with a target stat of 0 and duration of 1.1)?

@AdrienLeGuillou thank you for the helpful tip on vapply, and also your edgelist version of the code above seems to be working on first pass.

**Comment by smjenness** [comment]
> Both cases are dealing with migration and no desired ties between groups. Is there any chance you could comment on the relative merit of these approaches (offset with -Inf coef in formation and hard-coded removal of ties on migration VS. nodemix terms in formation and dissolution with a target stat of 0 and duration of 1.1)?

I haven't actually used the constraint method myself, so if that's working for you I would go with that. @chad-klumb may have further thoughts here.

**Comment by chad-klumb** [comment]
Even with duration exactly = 1 no finite number of proposals is guaranteed to remove all the hetero ties, so if no hetero ties is intended to be a constraint then I would go with manual removal.

**Comment by ThomasKraft** [comment]
> Even with duration exactly = 1 no finite number of proposals is guaranteed to remove all the hetero ties, so if no hetero ties is intended to be a constraint then I would go with manual removal.

This is exactly what I experienced when using `nodematch` or `nodemix` with target stats setting 100% or 0 ties. Hence why I moved towards the block diagonal formation constraint + manual removal. I have tested this very extensively since to ensure that it is behaving as desired and it works. I'm in the final phase of the project now and am still thinking it might be useful to add an abridged version of my migration module to your gallery for others to use as a template.
